src/librbd/crypto/FlattenRequest.cc

```diff
@@ -0,0 +1,248 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "FlattenRequest.h"
+
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/AsyncObjectThrottle.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/ObjectMap.h"
+#include "librbd/Utils.h"
+#include "librbd/crypto/CryptoInterface.h"
+#include "librbd/crypto/EncryptionFormat.h"
+#include "librbd/io/Utils.h"
+#include "librbd/operation/MetadataRemoveRequest.h"
+#include "osdc/Striper.h"
+#include <boost/lambda/bind.hpp>
+#include <boost/lambda/construct.hpp>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::crypto::FlattenRequest: " << this \
+                           << " " << __func__ << ": "
+
+namespace librbd {
+namespace crypto {
+
+using librbd::util::create_context_callback;
+
+template <typename I>
+class C_FlattenObject : public C_AsyncObjectThrottle<I> {
+public:
+  C_FlattenObject(AsyncObjectThrottle<I> &throttle, I *image_ctx,
+                  IOContext io_context, uint64_t object_no)
+    : C_AsyncObjectThrottle<I>(throttle, *image_ctx), m_io_context(io_context),
+      m_object_no(object_no) {
+  }
+
+  int send() override {
+    if (this->m_image_ctx.object_map != nullptr &&
+        !this->m_image_ctx.object_map->object_may_not_exist(m_object_no)) {
+      // can skip because the object already exists
+      return 1;
+    }
+
+    if (!io::util::trigger_copyup(
+            &this->m_image_ctx, m_object_no, m_io_context, this)) {
+      // stop early if the parent went away - it just means
+      // another flatten finished first or the image was resized
+      return 1;
+    }
+
+    return 0;
+  }
+
+private:
+  IOContext m_io_context;
+  uint64_t m_object_no;
+};
+
+template <typename I>
+FlattenRequest<I>::FlattenRequest(
+        I* image_ctx, Context* on_finish) : m_image_ctx(image_ctx),
+                                            m_raw_image_ctx(nullptr),
+                                            m_on_finish(on_finish) {
+}
+
+template <typename I>
+void FlattenRequest<I>::send() {
+  m_image_ctx->image_lock.lock_shared();
+  bool is_encryption_loaded = m_image_ctx->encryption_format.get() != nullptr;
+  m_image_ctx->image_lock.unlock_shared();
+  if (!is_encryption_loaded) {
+    // no encryption
+    finish(0);
+    return;
+  }
+
+  m_raw_image_ctx = I::create(m_image_ctx->name, m_image_ctx->id, nullptr,
+                              m_image_ctx->data_ctx, false);
+
+  auto ctx = create_context_callback<
+          FlattenRequest<I>, &FlattenRequest<I>::handle_create_raw_ctx>(this);
+  m_raw_image_ctx->state->open(0, ctx);
```

Comment:
You mean by adding a new object read flag to skip the crypto layers (object and image, which BTW has another PR to remove)?

Reply:
Yes, something like that -- although I must confess that I haven't thought through the details.  Even if there is a snag or two in there, it still feels like it would end up much better than opening the image.

> object and image, which BTW has another PR to remove

To remove what?

Reply:
> Yes, something like that -- although I must confess that I haven't thought through the details. Even if there is a snag or two in there, it still feels like it would end up much better than opening the image a second time.

So I went over the code trying to craft a solution in this fashion, but it is not obvious to me what's the best way of doing it.
Would appreciate your thoughts on how to do this more precisely.

Here's the flow I'm seeing:

Assume running "flatten" on an ImageCtx with encryption loaded.
If the image is crypto formatted everything works as the header is already present on the child, so we're just left with copying-up the data, which already works.
The problem is the case where the image is not crypto formatted, but rather one of its ancestors is crypto-formatted.
In that case, most chances are that the LUKS header is not present on the child, and we need to copy it up from its ancestor.
The flow in master is as flows:
1. `librbd::operation::FlattenRequest` is iterating all child objects and calling `io::util::trigger_copyup` on each one.
2. ``io::util::trigger_copyup` is calling the object core layer directly (`ObjectWriteRequest`), doing a zero-byte write to trigger a copyup.
3. ObjectWriteRequest computes parent extents using `io::util::extent_to_file`, which actually maps to negative images offsets (e.g. because object #0 on the child maps to image offset 0, but then we `remap_extents` and decrease the LUKS header offset from the image extents offsets).
4. `AbstractObjectWriteRequest<I>::copyup()` starts `librbd::io::CopyupRequest` with the negative parent extents from above.
5. `CopyupRequest<I>::read_from_parent` is creating an image dispatch read using the negative parent extents.
6. No `clip_io` will be called to find that the given extents are actually invalid (negative offsets).
7. Not sure how other image dispatch layers will react to this negative offsets.
8. The CORE image dispatch layer (`librbd::io::ImageReadRequest`) will call `librbd::io::util::file_to_extents` on the given extents, which will translate back to valid object extents which correspond to the LUKS header.
9. The spawned object requests above will go through crypto object dispatch layer - this should be avoided.
10. If the object does not exist on the parent (e.g. if the parent itself was not crypto-formatted, but rather some other ancestor), then `ObjectReadRequest<I>::read_parent()` will call `io::util::read_parent`, which again calls `remap_extents`, and gets us back to item 6 above.

Maybe we can keep the negative offsets as a hack to get over the `remap_extents`, but we're still need to skip the crypto object dispatch layer.
But probably a better idea is to avoid the negative offsets all-together (where? in `FlattenRequest`? `ObjectRequest`?)
The point in code where we detect this, also needs to start propagation of "skip object crypto layer" information.
This propagation needs to go through all the above steps (image dispatch layers, object dispatch layers + ``io::util::read_parent`).

Reply:
Finally decided on another solution which is to shut-down crypto after flattening the image data, which then allows flattening the header + restoring the LUKS magic.